app/scripts/controllers/permissions/index.js

```diff
@@ -188,33 +213,52 @@ export class PermissionsController {
    */
   async legacyExposeAccounts (origin, accounts) {
 
-    const permissions = {
-      eth_accounts: {},
+    // accounts are validated by finalizePermissionsRequest
+    if (typeof origin !== 'string' || !origin.length) {
+      throw new Error('Must provide non-empty string origin.')
+    }
+
+    const existingAccounts = await this.getAccounts(origin)
+
+    if (existingAccounts.length > 0) {
+      throw new Error(
+        'May not call legacyExposeAccounts on origin with exposed accounts.'
+      )
     }
 
-    await this.finalizePermissionsRequest(permissions, accounts)
+    const permissions = await this.finalizePermissionsRequest(
+      { eth_accounts: {} }, accounts
+    )
 
-    let error
     try {
+
       await new Promise((resolve, reject) => {
-        this.permissions.grantNewPermissions(origin, permissions, {}, (err) => (err ? resolve() : reject(err)))
+        this.permissions.grantNewPermissions(
+          origin, permissions, {}, _end
+        )
+
+        function _end (err) {
```

Comment:
This function is basically what `pify` does - we could use that here instead of creating a Promise constructor wrapper

Reply:
Good point, I'll give that a shot.

Reply:
It didn't immediately work (probably because of my own obtuseness?), so I have decided to give up on this effort for now.